state/src/skip_list.rs

```diff
@@ -0,0 +1,771 @@
+        Ok(Self {
+            trie,
+            level,
+            next_key: start_value
+                .next_value
+                .map(|t| SkipListKey::from_cow(account, level, t)),
+        })
+    }
+}
+
+impl<T, D> Iterator for SkipListIterator<'_, T, D>
+where
+    D: DB,
+    T: Clone + Listable + Serialize + DeserializeOwned + 'static,
+{
+    type Item = Result<T, TrieError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let key = self.next_key.take()?;
+        let value = match key.trie_value(self.trie).transpose()? {
+            Ok(value) => value,
+            Err(e) => {
+                return Some(Err(e));
+            }
+        };
+        let next_key = value
+            .next_value
+            .map(|t| SkipListKey::from_cow(key.account, self.level, t));
+        self.next_key = next_key;
+        Some(Ok(key.value.into_owned()))
+    }
+}
+
+type SkipListPair<T> = (SkipListKey<'static, T>, SkipListValue<'static, T>);
+
+fn search_item<T, D>(
+    account: AccountAddress,
+    max_levels: u32,
+    first_value: Cow<'static, T>,
+    element: &T,
+    trie: &EthTrie<D>,
+) -> Result<Vec<Option<SkipListPair<T>>>, TrieError>
+where
+    D: DB,
+    T: Clone + Listable + Ord + Serialize + DeserializeOwned + 'static,
+{
+    let mut prevs: Vec<Option<(SkipListKey<'static, T>, SkipListValue<'static, T>)>> =
+        vec![None; (max_levels + 1) as usize];
+
+    // If new element is before the first element then there is nothing to do.
+    if element < first_value.as_ref() {
+        return Ok(prevs);
+    }
+
+    let mut current_key = SkipListKey::from_cow(account, max_levels, first_value);
+
+    // Loop invariant: by design, `current_key.value < element`
+    loop {
+        let current_value = current_key
+            .trie_value(trie)?
+            .unwrap_or_else(|| panic!("Failed to find trie value for key account={account} level={} max_level={max_levels}", current_key.level));
+        let Some(next_value) = current_value.next_value.as_ref() else {
+            prevs[current_key.level as usize] = Some((current_key.clone(), current_value));
+            let Some(lower_level) = current_key.level.checked_sub(1) else {
+                return Ok(prevs);
+            };
+            current_key.level = lower_level;
+            continue;
+        };
+
+        // new element is after the next value too, keep going
+        if next_value.as_ref() < element {
+            current_key.value = current_value.next_value.expect("next_value is some");
+            continue;
+        }
+
+        // New element is between current element and next element
+        prevs[current_key.level as usize] = Some((current_key.clone(), current_value));
+        let Some(lower_level) = current_key.level.checked_sub(1) else {
+            return Ok(prevs);
+        };
+        current_key.level = lower_level;
+        continue;
+    }
+}
+
+fn pick_insert_height<R: Rng>(rng: &mut R) -> usize {
+    let mut result = 1;
```

Comment:
Why is the picked height is by default at least 1, though the base level is 0-indexed?